File: Sample/SampleOC/HomeViewController.m
//
//  HomeViewController.m
//  SampleOC
//
//  Created by KK on 2026/6/28.
//

#import "HomeViewController.h"

#import "LandscapeViewController.h"
#import "UserInfoViewController.h"

@import OrientationTransitionKit;

@interface HomeViewController () <OTKTransitionFromContextProvider>
@property (nonatomic, strong) UIView *playerContainerView;
@property (nonatomic, strong) UIImageView *playerView;
@property (nonatomic, strong) UIStackView *buttonStackView;
@property (nonatomic, strong) UIButton *landscapeButton;
@property (nonatomic, strong) UIButton *positionButton;
@property (nonatomic, strong) NSLayoutConstraint *playerContainerViewTopLayoutConstraint;
@property (nonatomic, strong) NSLayoutConstraint *playerContainerViewCenterLayoutConstraint;
@property (nonatomic, strong) OTKTransitionCoordinator *orientationTransitionCoordinator;
@property (nonatomic, strong) LandscapeViewController *landscapeViewController;
@property (nonatomic, copy) void (^pendingExitFullscreenTask)(void);
@property (nonatomic, assign) BOOL shouldHideHomeIndicator;
@end

@implementation HomeViewController

- (void)viewDidLoad {
    [super viewDidLoad];

    _playerContainerView = [UIView new];
    _playerContainerView.backgroundColor = UIColor.blackColor;
    _playerContainerView.translatesAutoresizingMaskIntoConstraints = NO;

    _playerView = [UIImageView new];
    _playerView.backgroundColor = UIColor.orangeColor;
    _playerView.autoresizingMask = UIViewAutoresizingFlexibleWidth | UIViewAutoresizingFlexibleHeight;

    _buttonStackView = [UIStackView new];
    _buttonStackView.translatesAutoresizingMaskIntoConstraints = NO;
    _buttonStackView.axis = UILayoutConstraintAxisHorizontal;
    _buttonStackView.spacing = 20;

    UIButtonConfiguration *positionButtonConfiguration = [UIButtonConfiguration filledButtonConfiguration];
    positionButtonConfiguration.contentInsets = NSDirectionalEdgeInsetsMake(5, 10, 5, 10);
    positionButtonConfiguration.cornerStyle = UIButtonConfigurationCornerStyleCapsule;
    positionButtonConfiguration.attributedTitle = [[NSAttributedString alloc] initWithString:@"顶部" attributes:@{NSFontAttributeName: [UIFont systemFontOfSize:18 weight:UIFontWeightMedium]}];
    _positionButton = [UIButton buttonWithConfiguration:positionButtonConfiguration primaryAction:nil];
    [_positionButton addTarget:self action:@selector(positionButtonTapped:) forControlEvents:UIControlEventTouchUpInside];

    UIButtonConfiguration *landscapeButtonConfiguration = [UIButtonConfiguration filledButtonConfiguration];
    landscapeButtonConfiguration.contentInsets = NSDirectionalEdgeInsetsMake(5, 10, 5, 10);
    landscapeButtonConfiguration.cornerStyle = UIButtonConfigurationCornerStyleCapsule;
    landscapeButtonConfiguration.attributedTitle = [[NSAttributedString alloc] initWithString:@"全屏" attributes:@{NSFontAttributeName: [UIFont systemFontOfSize:18 weight:UIFontWeightMedium]}];
    _landscapeButton = [UIButton buttonWithConfiguration:landscapeButtonConfiguration primaryAction:nil];
    _landscapeButton.translatesAutoresizingMaskIntoConstraints = NO;
    [_landscapeButton addTarget:self action:@selector(landscapeButtonTapped) forControlEvents:UIControlEventTouchUpInside];

    [self.view addSubview:_playerContainerView];
    [_playerContainerView addSubview:_playerView];
    [self.view addSubview:_buttonStackView];
    [_buttonStackView addArrangedSubview:_positionButton];
    [_buttonStackView addArrangedSubview:_landscapeButton];

    _playerContainerViewTopLayoutConstraint = [_playerContainerView.topAnchor constraintEqualToAnchor:self.view.topAnchor];
    _playerContainerViewCenterLayoutConstraint = [_playerContainerView.centerYAnchor constraintEqualToAnchor:self.view.centerYAnchor constant:-50];
    [NSLayoutConstraint activateConstraints:@[
        [_playerContainerView.leadingAnchor constraintEqualToAnchor:self.view.leadingAnchor],
        [_playerContainerView.trailingAnchor constraintEqualToAnchor:self.view.trailingAnchor],
        _playerContainerViewCenterLayoutConstraint,
        [_playerContainerView.heightAnchor constraintEqualToAnchor:_playerContainerView.widthAnchor multiplier:3.0 / 4.0],

        [_buttonStackView.topAnchor constraintEqualToAnchor:_playerContainerView.bottomAnchor constant:30],
        [_buttonStackView.centerXAnchor constraintEqualToAnchor:self.view.centerXAnchor],
    ]];

    UILabel *testLabel = [UILabel new];
    testLabel.font = [UIFont systemFontOfSize:30 weight:UIFontWeightMedium];
    testLabel.text = @"Test";
    testLabel.translatesAutoresizingMaskIntoConstraints = NO;
    [_playerView addSubview:testLabel];
    [testLabel.centerXAnchor constraintEqualToAnchor:_playerView.centerXAnchor].active = YES;
    [testLabel.centerYAnchor constraintEqualToAnchor:_playerView.centerYAnchor].active = YES;
}

- (void)viewDidLayoutSubviews {
    [super viewDidLayoutSubviews];

    if (self.playerView.superview == self.playerContainerView) {
        self.playerView.frame = self.playerContainerView.bounds;
    }
}

- (BOOL)prefersHomeIndicatorAutoHidden {
    return self.shouldHideHomeIndicator;
}

- (void)positionButtonTapped:(UIButton *)sender {
    sender.selected = !sender.selected;

    UIButtonConfiguration *configuration = [UIButtonConfiguration filledButtonConfiguration];
    configuration.contentInsets = NSDirectionalEdgeInsetsMake(5, 10, 5, 10);
    configuration.cornerStyle = UIButtonConfigurationCornerStyleCapsule;

    self.playerContainerViewCenterLayoutConstraint.active = NO;
    self.playerContainerViewTopLayoutConstraint.active = NO;

    if (sender.selected) {
        configuration.attributedTitle = [[NSAttributedString alloc] initWithString:@"中心" attributes:@{NSFontAttributeName: [UIFont systemFontOfSize:18 weight:UIFontWeightMedium]}];
        self.playerContainerViewTopLayoutConstraint.active = YES;
    } else {
        configuration.attributedTitle = [[NSAttributedString alloc] initWithString:@"顶部" attributes:@{NSFontAttributeName: [UIFont systemFontOfSize:18 weight:UIFontWeightMedium]}];
        self.playerContainerViewCenterLayoutConstraint.active = YES;
    }
    sender.configuration = configuration;

    [UIView animateWithDuration:0.25 animations:^{
        [self.view layoutIfNeeded];
    }];
}

- (void)landscapeButtonTapped {
    if (self.landscapeViewController != nil) {
        return;
    }

    LandscapeViewController *landscapeViewController = [LandscapeViewController new];
    landscapeViewController.transitionContentView = self.playerView;

    __weak typeof(self) weakSelf = self;
    landscapeViewController.didTapUserInfoHandler = ^(LandscapeViewController *viewController) {
        weakSelf.pendingExitFullscreenTask = ^{
            [weakSelf showUserInfo];
        };
        [viewController dismissViewControllerAnimated:YES completion:nil];
    };

    OTKDefaultTransitionAnimationProvider *animationProvider = [[OTKDefaultTransitionAnimationProvider alloc] initWithDuration:0.25];
    OTKTransitionCoordinator *transitionCoordinator = [[OTKTransitionCoordinator alloc] initFromContextProvider:self toContextProvider:landscapeViewController fromInterfaceOrientation:UIInterfaceOrientationPortrait toInterfaceOrientation:UIInterfaceOrientationLandscapeRight animationProvider:animationProvider];

    self.landscapeViewController = landscapeViewController;
    self.orientationTransitionCoordinator = transitionCoordinator;

    landscapeViewController.modalPresentationStyle = UIModalPresentationFullScreen;
    landscapeViewController.transitioningDelegate = transitionCoordinator;

    [self presentViewController:landscapeViewController animated:YES completion:nil];
}

- (void)showUserInfo {
    [self.navigationController pushViewController:[UserInfoViewController new] animated:YES];
}

- (void)setHomeIndicatorHidden:(BOOL)hidden {
    self.shouldHideHomeIndicator = hidden;
    [self setNeedsUpdateOfHomeIndicatorAutoHidden];
}

- (UIViewController *)transitionFromContextProviderViewController:(id<OTKTransitionFromContextProvider>)contextProvider {
    return self;
}

- (CGRect)transitionFromContextProviderTransitionFrame:(id<OTKTransitionFromContextProvider>)contextProvider in:(UIView *)containerView {
    return [containerView convertRect:self.playerContainerView.bounds fromView:self.playerContainerView];
}

- (void)transitionFromContextProviderPrepareTransitionView:(id<OTKTransitionFromContextProvider>)contextProvider transitionView:(UIView *)transitionView {
    [self movePlayerViewToContainerView:transitionView];
}

- (void)transitionFromContextProviderFinishTransitionView:(id<OTKTransitionFromContextProvider>)contextProvider {
    [self movePlayerViewToContainerView:self.playerContainerView];
}

- (void)transitionFromContextProviderTransitionWillEnter:(id<OTKTransitionFromContextProvider>)contextProvider to:(id<OTKTransitionToContextProvider>)toContextProvider {
    [self setHomeIndicatorHidden:YES];
}

- (void)transitionFromContextProviderTransitionDidEnter:(id<OTKTransitionFromContextProvider>)contextProvider to:(id<OTKTransitionToContextProvider>)toContextProvider {
    [self setHomeIndicatorHidden:NO];
}

- (void)transitionFromContextProviderTransitionWillExit:(id<OTKTransitionFromContextProvider>)contextProvider to:(id<OTKTransitionToContextProvider>)toContextProvider {
    [self setHomeIndicatorHidden:YES];
}

- (void)transitionFromContextProviderTransitionDidExit:(id<OTKTransitionFromContextProvider>)contextProvider to:(id<OTKTransitionToContextProvider>)toContextProvider {
    [self setHomeIndicatorHidden:NO];
    self.orientationTransitionCoordinator = nil;
    self.landscapeViewController = nil;

    if (self.pendingExitFullscreenTask != nil) {
        self.pendingExitFullscreenTask();
        self.pendingExitFullscreenTask = nil;
    }
}

- (void)movePlayerViewToContainerView:(UIView *)containerView {
    [self.playerView removeFromSuperview];
    self.playerView.transform = CGAffineTransformIdentity;
    self.playerView.frame = containerView.bounds;
    self.playerView.autoresizingMask = UIViewAutoresizingFlexibleWidth | UIViewAutoresizingFlexibleHeight;
    [containerView addSubview:self.playerView];
}

@end

File: Sample/Sample/HomeViewController.swift
//
//  HomeViewController.swift
//  OrientationTransitionKitSample
//
//  Created by KK on 2026/6/28.
//

import OrientationTransitionKit
import UIKit

class HomeViewController: BaseViewController {
    private let playerContainerView = UIView()
    private let playerView = UIImageView()
    private let buttonStackView = UIStackView()
    private let landscapeButton = UIButton(configuration: .plain())
    private let systemRotationLandscapeButton = UIButton(configuration: .plain())
    private let positionButton = UIButton(configuration: .plain())

    private var playerContainerViewTopLayoutConstraint: NSLayoutConstraint!
    private var playerContainerViewCenterLayoutConstraint: NSLayoutConstraint!

    private var orientationTransitionCoordinator: TransitionCoordinator?
    private var systemRotationTransitionCoordinator: SystemRotationTransitionCoordinator?
    private var landscapeViewController: LandscapeViewController?
    private var pendingExitFullscreenTask: (() -> Void)?
    private var shouldHideHomeIndicator = false

    override func viewDidLoad() {
        super.viewDidLoad()

        playerContainerView.backgroundColor = .black
        playerContainerView.clipsToBounds = true
        playerContainerView.layer.cornerRadius = 20
        playerContainerView.translatesAutoresizingMaskIntoConstraints = false

        playerView.backgroundColor = .orange
//        playerView.autoresizingMask = [.flexibleWidth, .flexibleHeight]
        playerView.translatesAutoresizingMaskIntoConstraints = false

        buttonStackView.translatesAutoresizingMaskIntoConstraints = false
        buttonStackView.axis = .horizontal
        buttonStackView.spacing = 20

        var positionButtonConfiguration = UIButton.Configuration.filled()
        positionButtonConfiguration.contentInsets = NSDirectionalEdgeInsets(top: 5, leading: 10, bottom: 5, trailing: 10)
        positionButtonConfiguration.cornerStyle = .capsule
        positionButtonConfiguration.attributedTitle = AttributedString(
            "顶部",
            attributes: AttributeContainer([
                .font: UIFont.systemFont(ofSize: 18, weight: .medium),
            ])
        )
        positionButton.configuration = positionButtonConfiguration
        positionButton.addTarget(self, action: #selector(positionButtonTapped(_:)), for: .touchUpInside)

        var landscapeButtonConfiguration = UIButton.Configuration.filled()
        landscapeButtonConfiguration.contentInsets = NSDirectionalEdgeInsets(top: 5, leading: 10, bottom: 5, trailing: 10)
        landscapeButtonConfiguration.cornerStyle = .capsule
        landscapeButtonConfiguration.attributedTitle = AttributedString(
            "全屏",
            attributes: AttributeContainer([
                .font: UIFont.systemFont(ofSize: 18, weight: .medium),
            ])
        )
        landscapeButton.configuration = landscapeButtonConfiguration
        landscapeButton.translatesAutoresizingMaskIntoConstraints = false
        landscapeButton.addTarget(self, action: #selector(landscapeButtonTapped), for: .touchUpInside)

        var systemRotationLandscapeButtonConfiguration = UIButton.Configuration.filled()
        systemRotationLandscapeButtonConfiguration.contentInsets = NSDirectionalEdgeInsets(top: 5, leading: 10, bottom: 5, trailing: 10)
        systemRotationLandscapeButtonConfiguration.cornerStyle = .capsule
        systemRotationLandscapeButtonConfiguration.attributedTitle = AttributedString(
            "全屏2",
            attributes: AttributeContainer([
                .font: UIFont.systemFont(ofSize: 18, weight: .medium),
            ])
        )
        systemRotationLandscapeButton.configuration = systemRotationLandscapeButtonConfiguration
        systemRotationLandscapeButton.translatesAutoresizingMaskIntoConstraints = false
        systemRotationLandscapeButton.addTarget(self, action: #selector(systemRotationLandscapeButtonTapped), for: .touchUpInside)

        view.addSubview(playerContainerView)
        playerContainerView.addSubview(playerView)
        view.addSubview(buttonStackView)
        buttonStackView.addArrangedSubview(positionButton)
        buttonStackView.addArrangedSubview(landscapeButton)
        buttonStackView.addArrangedSubview(systemRotationLandscapeButton)

        playerContainerViewTopLayoutConstraint = playerContainerView.topAnchor.constraint(equalTo: view.topAnchor)
        playerContainerViewCenterLayoutConstraint = playerContainerView.centerYAnchor.constraint(equalTo: view.centerYAnchor, constant: -50)
        NSLayoutConstraint.activate([
            playerContainerView.leadingAnchor.constraint(equalTo: view.leadingAnchor),
            playerContainerView.trailingAnchor.constraint(equalTo: view.trailingAnchor),
            playerContainerViewCenterLayoutConstraint,
            playerContainerView.heightAnchor.constraint(equalTo: playerContainerView.widthAnchor, multiplier: 3.0 / 4.0),

            playerView.leadingAnchor.constraint(equalTo: playerContainerView.leadingAnchor),
            playerView.trailingAnchor.constraint(equalTo: playerContainerView.trailingAnchor),
            playerView.topAnchor.constraint(equalTo: playerContainerView.topAnchor),
            playerView.bottomAnchor.constraint(equalTo: playerContainerView.bottomAnchor),

            buttonStackView.topAnchor.constraint(equalTo: playerContainerView.bottomAnchor, constant: 30),
            buttonStackView.centerXAnchor.constraint(equalTo: view.centerXAnchor),
        ])

        let testLabel = UILabel()
        testLabel.font = UIFont.systemFont(ofSize: 40, weight: .medium)
        testLabel.text = "Test"
        playerView.addSubview(testLabel)
        testLabel.translatesAutoresizingMaskIntoConstraints = false
        testLabel.centerXAnchor.constraint(equalTo: playerView.centerXAnchor).isActive = true
        testLabel.centerYAnchor.constraint(equalTo: playerView.centerYAnchor).isActive = true
    }

    override func viewDidLayoutSubviews() {
        super.viewDidLayoutSubviews()

//        if playerView.superview === playerContainerView {
//            playerView.frame = playerContainerView.bounds
//        }
    }

    override var prefersHomeIndicatorAutoHidden: Bool {
        shouldHideHomeIndicator
    }

    @objc private func positionButtonTapped(_ sender: UIButton) {
        sender.isSelected.toggle()

        var configuration = UIButton.Configuration.filled()
        configuration.contentInsets = NSDirectionalEdgeInsets(top: 5, leading: 10, bottom: 5, trailing: 10)
        configuration.cornerStyle = .capsule
        playerContainerViewCenterLayoutConstraint.isActive = false
        playerContainerViewTopLayoutConstraint.isActive = false

        if sender.isSelected {
            configuration.attributedSubtitle = AttributedString(
                "中心",
                attributes: AttributeContainer([
                    .font: UIFont.systemFont(ofSize: 18, weight: .medium),
                ])
            )
            playerContainerViewTopLayoutConstraint.isActive = true
        } else {
            configuration.attributedSubtitle = AttributedString(
                "顶部",
                attributes: AttributeContainer([
                    .font: UIFont.systemFont(ofSize: 18, weight: .medium),
                ])
            )
            playerContainerViewCenterLayoutConstraint.isActive = true
        }
        sender.configuration = configuration

        UIView.animate(withDuration: 0.25) {
            self.view.layoutIfNeeded()
        }
    }

    @objc private func landscapeButtonTapped() {
        guard landscapeViewController == nil else {
            return
        }

        let landscapeViewController = LandscapeViewController()
        landscapeViewController.transitionContentView = playerView
        landscapeViewController.didTapUserInfoHandler = { [weak self] viewController in
            self?.pendingExitFullscreenTask = {
                self?.showUserInfo()
            }

            viewController.dismiss(animated: true)
        }

        let animationProvider = DefaultTransitionAnimationProvider()
        let transitionCoordinator = TransitionCoordinator(
            fromContextProvider: self,
            toContextProvider: landscapeViewController,
            fromInterfaceOrientation: .portrait,
            toInterfaceOrientation: .landscapeRight,
            animationProvider: animationProvider
        )

        self.landscapeViewController = landscapeViewController
        orientationTransitionCoordinator = transitionCoordinator

        landscapeViewController.modalPresentationStyle = .fullScreen
        landscapeViewController.transitioningDelegate = transitionCoordinator

        present(landscapeViewController, animated: true)
    }

    @objc private func systemRotationLandscapeButtonTapped() {
        guard landscapeViewController == nil else {
            return
        }

        let landscapeViewController = LandscapeViewController()
        landscapeViewController.transitionContentView = playerView
        landscapeViewController.didTapCloseHandler = { [weak self] _ in
            self?.systemRotationTransitionCoordinator?.dismiss()
        }
        landscapeViewController.didTapUserInfoHandler = { [weak self] _ in
            self?.pendingExitFullscreenTask = {
                self?.showUserInfo()
            }
            self?.systemRotationTransitionCoordinator?.dismiss()
        }

        let transitionCoordinator = SystemRotationTransitionCoordinator(
            fromContextProvider: self,
            toContextProvider: landscapeViewController,
            fromInterfaceOrientation: .portrait,
            toInterfaceOrientation: .landscapeRight
        )

        self.landscapeViewController = landscapeViewController
        systemRotationTransitionCoordinator = transitionCoordinator
        transitionCoordinator.present()
    }

    private func showUserInfo() {
        navigationController?.pushViewController(UserInfoViewController(), animated: true)
    }

    private func setHomeIndicatorHidden(_ isHidden: Bool) {
        shouldHideHomeIndicator = isHidden
        setNeedsUpdateOfHomeIndicatorAutoHidden()
    }
}

extension HomeViewController: TransitionFromContextProvider {
    func transitionFromContextProviderViewController(_ contextProvider: TransitionFromContextProvider) -> UIViewController {
        self
    }

    func transitionFromContextProviderTransitionFrame(_ contextProvider: TransitionFromContextProvider, in containerView: UIView) -> CGRect {
        containerView.convert(playerContainerView.bounds, from: playerContainerView)
    }

    func transitionFromContextProviderPrepareTransitionView(_ contextProvider: TransitionFromContextProvider, transitionView: UIView) {
        movePlayerView(to: transitionView)
    }

    func transitionFromContextProviderAnimateAlongsideTransition(_ contextProvider: TransitionFromContextProvider, transitionView: UIView, animator: UIViewImplicitlyAnimating) {
        transitionView.clipsToBounds = true
        transitionView.layer.cornerRadius = 20
        animator.addAnimations? {
            transitionView.layer.cornerRadius = 0
        }
    }

    func transitionFromContextProviderFinishTransitionView(_ contextProvider: TransitionFromContextProvider) {
        movePlayerView(to: playerContainerView)
    }

    func transitionFromContextProviderMakePresentingSnapshotView(
        _ contextProvider: TransitionFromContextProvider,
        afterScreenUpdates: Bool
    ) -> UIView? {
        playerView.isHidden = true
        defer {
            playerView.isHidden = false
        }

        // snapshotView may reuse the previous render cache, so the temporary hidden playerView can still appear in it.
        view.layoutIfNeeded()
        let renderer = UIGraphicsImageRenderer(bounds: view.bounds)
        let image = renderer.image { _ in
            view.drawHierarchy(in: view.bounds, afterScreenUpdates: true)
        }
        let snapshotView = UIImageView(image: image)
        snapshotView.frame = view.bounds
        return snapshotView
    }

    func transitionFromContextProviderTransitionWillEnter(_ contextProvider: TransitionFromContextProvider, to toContextProvider: TransitionToContextProvider) {
        setHomeIndicatorHidden(true)
    }

    func transitionFromContextProviderTransitionDidEnter(_ contextProvider: TransitionFromContextProvider, to toContextProvider: TransitionToContextProvider) {
        setHomeIndicatorHidden(false)
    }

    func transitionFromContextProviderTransitionWillExit(_ contextProvider: TransitionFromContextProvider, to toContextProvider: TransitionToContextProvider) {
        setHomeIndicatorHidden(true)
    }

    func transitionFromContextProviderTransitionDidExit(_ contextProvider: TransitionFromContextProvider, to toContextProvider: TransitionToContextProvider) {
        setHomeIndicatorHidden(false)
        orientationTransitionCoordinator = nil
        systemRotationTransitionCoordinator = nil
        landscapeViewController = nil

        pendingExitFullscreenTask?()
        pendingExitFullscreenTask = nil
    }

    private func movePlayerView(to containerView: UIView) {
        playerView.removeFromSuperview()
        playerView.transform = .identity
        playerView.frame = containerView.bounds
        playerView.translatesAutoresizingMaskIntoConstraints = false
//        playerView.autoresizingMask = [.flexibleWidth, .flexibleHeight]
        containerView.addSubview(playerView)
        NSLayoutConstraint.activate([
            playerView.leadingAnchor.constraint(equalTo: containerView.leadingAnchor),
            playerView.trailingAnchor.constraint(equalTo: containerView.trailingAnchor),
            playerView.topAnchor.constraint(equalTo: containerView.topAnchor),
            playerView.bottomAnchor.constraint(equalTo: containerView.bottomAnchor),
        ])
    }
}
